doc/manual/interfaces.rst

```diff
@@ -147,25 +147,28 @@ While this is starting to support more of the :ref:`indexing operations supporte
+Methods to implement                                                                                               Brief description
+===================================================================== ============================================ =======================================================================================
+:func:`size(A) <size>`                                                                                             Returns a tuple containing the dimensions of ``A``
+:func:`getindex(A, i::Int) <getindex>`                                                                             (if ``LinearFast``) Linear scalar indexing
+:func:`getindex(A, I::Vararg{Int, N}) <getindex>`                                                                  (if ``LinearSlow``, where ``N = ndims(A)``) N-dimensional scalar indexing
+:func:`setindex!(A, v, i::Int) <setindex!>`                                                                        (if ``LinearFast``) Scalar indexed assignment
+:func:`setindex!(A, v, I::Vararg{Int, N}) <setindex!>`                                                             (if ``LinearSlow``, where ``N = ndims(A)``) N-dimensional scalar indexed assignment
+**Optional methods**                                                  **Default definition**                       **Brief description**
+:func:`Base.linearindexing(Type) <Base.linearindexing>`               ``Base.LinearSlow()``                        Returns either ``Base.LinearFast()`` or ``Base.LinearSlow()``. See the description below.
+:func:`indices(A, d) <indices>`                                       ``1:size(A, d)``                             Returns the range of valid indices along dimension ``d``
+:func:`getindex(A, I...) <getindex>`                                  defined in terms of scalar :func:`getindex`  :ref:`Multidimensional and nonscalar indexing <man-array-indexing>`
+:func:`setindex!(A, I...) <setindex!>`                                defined in terms of scalar :func:`setindex!` :ref:`Multidimensional and nonscalar indexed assignment <man-array-indexing>`
+:func:`start`/:func:`next`/:func:`done`                               defined in terms of scalar :func:`getindex`  Iteration
+:func:`length(A) <length>`                                            ``prod(size(A))``                            Number of elements
+:func:`similar(A) <similar>`                                          ``similar(A, eltype(A), indices(A))``        Return a mutable array with the same shape and element type
+:func:`similar(A, ::Type{S}) <similar>`                               ``similar(A, S, indices(A))``                Return a mutable array with the same shape and the specified element type
+:func:`similar(A, inds::NTuple{UnitRange{Int}}) <similar>`            ``similar(A, eltype(A), inds)``              Return a mutable array with the same element type and the specified indices
```

Comment:
It's interesting to think about how non-scalar indexing will use this definition.  You'll probably want the output array to match the indices of the indexes :).  That is: `A[2, ZeroArray([3,2,1])]` should probably call `similar(A, 0:2)` and return a `ZeroArray([A[2, 3], A[2, 2], A[2, 1]])`.  Even more fascinating: this would be completely impossible if we didn't drop scalar dimensions.  I didn't see that coming, did you?

So then, what happens with `similar(::Array, ::UnitRange)`?  We're effectively extending the awkwardness that is the element-type/wrapper-type difficulties between `Array{Bool}` and `BitArray` to a *third* dimension: `OffsetArray{Bool}`.  Only this one is worse, since an `Array` can't emulate an `OffsetArray`. My heels aren't quite so firmly in the mud as they used to be, but I still think that this will make generic code harder to write.